src/main/scala/amba/apb/ToTL.scala

```diff
@@ -0,0 +1,83 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.amba.apb
+
+import Chisel._
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util.MaskGen
+
+case class APBToTLNode()(implicit valName: ValName) extends MixedAdapterNode(APBImp, TLImp)(
+  dFn = { case APBMasterPortParameters(masters) =>
+    TLClientPortParameters(clients = masters.map { m =>
+      TLClientParameters(name = m.name, nodePath = m.nodePath)
+    })
+  },
+  uFn = { mp => APBSlavePortParameters(
+    slaves = mp.managers.map { m =>
+      APBSlaveParameters(
+        address       = m.address,
+        resources     = m.resources,
+        regionType    = m.regionType,
+        executable    = m.executable,
+        nodePath      = m.nodePath,
+        //TODO why and how should I do this adjust for APB?
+        // Also, shouldn't I use supportsPutPartial because of PSTRB?
+        supportsWrite = m.supportsPutFull,//adjust(m.supportsPutFull),
+        supportsRead  = m.supportsGet)},//adjust(m.supportsGet))},
+    beatBytes = mp.beatBytes)
+  }) {
+  //TODO: Shouldn't there be a requirement that TL side is wider than APB side?
+}
+
+class APBToTL()(implicit p: Parameters) extends LazyModule
+{
+  val node = APBToTLNode()
+
+  lazy val module = new LazyModuleImp(this) {
+    (node.in zip node.out) foreach { case ((in, edgeIn), (out, edgeOut)) =>
+      val beatBytes = edgeOut.manager.beatBytes
+
+      //-----------------------------------------------
+      // Flow Control
+      out.a.valid := in.psel && in.penable
+      in.pready := in.psel & in.penable & out.d.valid
+      out.d.ready := in.psel && in.penable // No backpressure to APB Master
+
+      // Write
+      // Question -- do I want PutPartial here or PutFull?
+      // I was thinking PutPartial because of the pstrb masking.
+      out.a.bits.opcode  := Mux(in.pwrite, TLMessages.PutPartialData, TLMessages.Get)
+      out.a.bits.param   := UInt(0)
+      out.a.bits.size    := UInt(log2Ceil(in.params.dataBits/8))
+      out.a.bits.source  := UInt(0)
+      out.a.bits.address := in.paddr
+      out.a.bits.data    := in.pwdata
+      val pstrbFill = beatBytes / (in.params.dataBits/8)
+      out.a.bits.mask    := MaskGen(out.a.bits.address, out.a.bits.size, beatBytes) | Cat(Seq.fill(pstrbFill)(in.pstrb).reverse)
```

Comment:
There are several things wrong with this logic.

First, the whole pstrbFill is pointless. out.manager.beatBytes must equal in.manager.beatBytes due to your (correct) Node rules. It then follows that beatBytes == in.params.dataBits/8.

Second, on a TileLink Get the mask must be the result you computed with MaskGen.

This is the correct line:
out.a.bits.mask := Mux(in.pwrite, in.pstrb, MaskGen(out.a.bits.address, out.a.bits.size, beatBytes))

Reply:
Actually, there is another problem I didn't notice on my first review pass: paddr. TileLink requests must be aligned to their size. I think you are correct to set size unilaterally to log2Ceil(beatBytes).U

However, this means you need to force alignment of the TileLink address:
out.a.bits.address := ~(~in.paddr | (beatBytes-1).U)
// The double negative here is to work around Chisel's broken implementation of widening ~x.

That also means that Gets are always full width:
out.a.bits.mask := Mux(in.pwrite, in.pstrb, ~0.U(beatBytes.W))

Reply:
For the address, should I be adding an assertion that it's aligned when it comes in? Or will TL assertions already check that?

Reply:
Oh I see above that you said I should force alignment. 

Shouldn't the user be required to do that? Othewise I'm not sure what the address means that I am receiving.

Reply:
For now I am adding an assertion that the alignment logic above does nothing.

Reply:
I think you should silently accept a misaligned address. If the bus is 4-byte wide, you might get a write with address = 6 and strb=0100. Converting that to address = 4 and strb=0100 is fine. In the other AMBA protocols there is a 'size' field, but APB does not have one. For AXI, the equivalent transformation would be to align the address and round the size up to the bus width. That does not change the amount written. It marginally changes the semantics of the operation, because the size was adjusted. However, for APB, you don't have a size field. Your adapter is just assuming all operations are bus-width and setting size to a constant, a decision I agree with. Thus you need the other half of the transformation: round the address down.